server/src/main/java/org/elasticsearch/index/query/DistanceFeatureQueryBuilder.java

```diff
@@ -0,0 +1,215 @@
+
+/**
+ * A query to boost scores based on their proximity to the given origin
+ * for date, date_nanos and geo_point field types
+ */
+public class DistanceFeatureQueryBuilder extends AbstractQueryBuilder<DistanceFeatureQueryBuilder> {
+    public static final String NAME = "distance_feature";
+
+    private static final ParseField FIELD_FIELD = new ParseField("field");
+    private static final ParseField ORIGIN_FIELD = new ParseField("origin");
+    private static final ParseField PIVOT_FIELD = new ParseField("pivot");
+
+    private final String field;
+    private final Origin origin;
+    private final String pivot;
+
+    private static final ConstructingObjectParser<DistanceFeatureQueryBuilder, Void> PARSER = new ConstructingObjectParser<>(
+        "distance_feature", false,
+        args -> new DistanceFeatureQueryBuilder((String) args[0], (Origin) args[1], (String) args[2])
+    );
+
+    static {
+        PARSER.declareString(constructorArg(), FIELD_FIELD);
+        // origin: number or string for date and date_nanos fields; string, array, object for geo fields
+        PARSER.declareField(constructorArg(), DistanceFeatureQueryBuilder.Origin::originFromXContent,
+            ORIGIN_FIELD, ObjectParser.ValueType.OBJECT_ARRAY_STRING_OR_NUMBER);
+        PARSER.declareString(constructorArg(), PIVOT_FIELD);
+        declareStandardFields(PARSER);
+    }
+
+    public DistanceFeatureQueryBuilder(String field, Origin origin, String pivot) {
+        this.field = Objects.requireNonNull(field);
+        this.origin = Objects.requireNonNull(origin);
+        this.pivot = Objects.requireNonNull(pivot);
+    }
+
+    public static DistanceFeatureQueryBuilder fromXContent(XContentParser parser) {
+        return PARSER.apply(parser, null);
+    }
+
+    @Override
+    protected void doXContent(XContentBuilder builder, Params params) throws IOException {
+        builder.startObject(NAME);
+        builder.field(FIELD_FIELD.getPreferredName(), field);
+        builder.field(ORIGIN_FIELD.getPreferredName(), origin.origin);
+        builder.field(PIVOT_FIELD.getPreferredName(), pivot);
+        printBoostAndQueryName(builder);
+        builder.endObject();
+    }
+
+    public DistanceFeatureQueryBuilder(StreamInput in) throws IOException {
+        super(in);
+        field = in.readString();
+        origin = new Origin(in);
+        pivot = in.readString();
+    }
+
+    @Override
+    protected void doWriteTo(StreamOutput out) throws IOException {
+        out.writeString(field);
+        origin.writeTo(out);
+        out.writeString(pivot);
+    }
+
+    @Override
+    public String getWriteableName() {
+        return NAME;
+    }
+
+    @Override
+    protected Query doToQuery(QueryShardContext context) throws IOException {
+        MappedFieldType fieldType = context.fieldMapper(field);
+        if (fieldType == null) {
+            throw new IllegalArgumentException("Can't run [" + NAME + "] query on unmapped fields!");
```

Comment:
our general policy is to be lenient when it comes to unmapped fields, to make cross-index search easier. I'd return a `MatchNoDocsQuery` instead.